### PR TITLE
feat(flow): added flow execute command

### LINKED ISSUE #249: sasjs flow
The official SAS Job Flow creator is a licensed product and not all clients will have it available.  So we need to create a rudimentary scheduler.  This will take a JSON file as input, and submit all the jobs sequentially, logging the results to an output CSV.

If a job errors, then the flow stops.

The input JSON will look like this:

```
{
  "name": "myAmazingFlow",
  "flows": {
    "firstflow": {
      "jobs": [
        {
          "location": "/Projects/job1"
        },
        {
          "location": "/Projects/job2"
        },
        {
          "location": "/Projects/job3"
        }
      ],
      "predecessors": []
    },
    "secondflow": {
      "jobs": [
        {
          "location": "/Projects/job11"
        }
      ],
      "predecessors": [
        "firstflow"
      ]
    },
    "anotherflow": {
      "jobs": [
        {
          "location": "/Public/job15"
        }
      ],
      "predecessors": [
        "firstflow",
        "secondflow"
      ]
    }
  },
    "yetanotherflow": {
      "jobs": [
        {
          "location": "/Public/job115"
        }
      ],
      "predecessors": []
    }
  }
}
```

This will contain a flow with 4 milestones:   

1. firstflow  (triggers 3 jobs at once, runs immediately as no predecessors
2. yetanotherflow (triggers 1 job, runs immediately as no predecesors)
3. secondflow (triggers as soon as firstflow is complete and ONLY if the jobs all finish with SUCCESS status)
4. anotherflow (triggers once firstflow and secondflow are complete, but ONLY if all finish with SUCCESS)

As soon as each job is complete, a CSV file is updated showing hte following attributes:

* jobLocation
* startTime
* endTime
* duration (formatted)
* logLocation

Syntax:

```
sasjs flow execute --source /local/flow.json --logFolder /local/log/folder --csvLocation /local/some.csv
```

### PR DESCRIPTION
**Pull request by YuryShkoda**
## Issue

closes #249 

## Intent

Create  a rudimentary scheduler that will take a JSON file as input, and submit all the jobs sequentially, logging the results to an output CSV.

## Implementation

1. Added new command.
2. Updated command utility.
3. Added unit tests.

## Checks

- [x] Code is formatted correctly (`npm run lint:fix`).
- [x] Any new functionality has been unit tested.
- [x] All unit tests are passing (`npm test`).
image
- [x] All CI checks are green.
- [x] JSDoc comments have been added or updated.
- [x] Reviewer is assigned.

### CODE REVIEW

File: src/utils/file-utils.js [hunk 1]
```diff
@@ -80,6 +80,26 @@ export async function readFile(fileName, debug = false, silent = false) {
   })
 }
 
+export async function writeFile(fileName, data, encoding = 'utf8') {
```
Comment (krishna-acondy) [inline src/utils/file-utils.js:83]: I believe we already had a `createFile` function that does exactly this.

File: src/commands/flow/execute.ts [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+
+    let sourceConfig = await readFile(source)
+
+    try {
+      sourceConfig = JSON.parse(sourceConfig)
+    } catch (error) {
+      throw `Invalid json file.`
+    }
+
+    let flows = sourceConfig.flows
```
Comment (krishna-acondy) [inline src/commands/flow/execute.ts:63]: I'd suggest having slightly stricter validation for each `flow` object.
We can do that by having a `Flow` class with a static `parse` method which throws errors on format mismatch.
If the JSON is valid, it'll return a new `Flow` object.

Comment (krishna-acondy) [reply src/commands/flow/execute.ts:63]: The class can have properties such as `predecessors` and `successors` which will give you auto-completion in the IDE as well.

File: package.json [hunk 3]
```diff
@@ -41,11 +41,13 @@
     "access": "public"
   },
   "dependencies": {
-    "@sasjs/adapter": "^1.18.7",
+    "@babel/preset-typescript": "^7.12.7",
```
Comment (krishna-acondy) [inline package.json:44]: This needs to be a `devDependency`.

File: package.json [hunk 4]
```diff
@@ -59,13 +61,14 @@
     "prompt": "^1.0.0",
     "rimraf": "^3.0.2",
     "shelljs": "^0.8.4",
+    "ts-jest": "^26.4.4",
```
Comment (krishna-acondy) [inline package.json:64]: This needs to be a `devDependency`.

File: src/commands/flow/execute.ts (hunk 2 above)
Comment (YuryShkoda) [reply src/commands/flow/execute.ts:62]: Good idea, I'll create a ticket for refactoring

**Review by allanbowe — APPROVED** [review]

**Pull request merged**